### PR TITLE
Issue/3209 modules v2 compilation tests

### LINKED ISSUE #3209: modules v2 pypi loader test: project compile
1. Check what exists, extend with assertions 

Add tests for compilation of a project. Edge cases to consider:
- v1 depends on v2
  - v2 was preinstalled
  - v2 was not preinstalled => installed on the fly
- v1 depends on v1
- v2 depends on other v2
  - other v2 was preinstalled
  - other v2 was not preinstalled => fail
- v2 depends on v1 => fail
- ...?

### PR DESCRIPTION
**Pull request by sanderr**
# Description

Added compilation test for modules v2. This test should cover compilation-specific behavior that has not been covered yet by the module and project loader tests.

closes #3209

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [x] ~~End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )~~

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: tests/compiler/test_basics.py [hunk 1]
```diff
@@ -212,3 +218,116 @@ def test_value_set_twice(snippetcompiler):
...
+    # install module as v2 (on top of v1) and verify compile
+    v2_module_path: str = os.path.join(str(tmpdir), test_module)
+    module_from_template(
+        v2_template_path,
+        v2_module_path,
+        new_name=test_module,
+        new_content_init_cf=test_module_model,
+        new_content_init_py=test_module_plugin_contents(2),
+        install=True,
+        editable=True,
```
Comment (wouterdb) [inline tests/compiler/test_basics.py:318]: I would add a comment that this is editable 

```suggestion
        editable=True,  # ! this is editable for the next test step
```

File: tests/compiler/test_basics.py [hunk 2]
```diff
@@ -212,3 +218,116 @@ def test_value_set_twice(snippetcompiler):
...
+        fh.write(test_module_plugin_contents(3))
+    verify_compile(3)
+
+    # verify editable mode for model
+    with open(os.path.join(v2_module_path, "model", "_init.cf"), "w") as fh:
+        fh.write("value = 4")
+    # can't just verify_compile(4) because the plugin is still at 3 and promoting it to 4 would not test changes in the model
+    with pytest.raises(DoubleSetException) as excinfo:
+        verify_compile(3)
+    assert excinfo.value.newvalue == 4
```
Comment (wouterdb) [inline tests/compiler/test_basics.py:333]: Strictly speaking either the old or new value could be 4, but we can fix that if it every breaks

**Review by wouterdb — APPROVED** [review]

File: tests/compiler/test_basics.py (hunk 2 above)
Comment (sanderr) [reply tests/compiler/test_basics.py:333]: Indeed, I was thinking the same. This should be stable unless we make explicit changes to the compile flow, or if there is some nondeterministic behavior we're not aware of. In the first case it's easy to fix, in the second case we'd want to make an attempt to make the behavior deterministic I think.

**Review by arnaudsjs — APPROVED** [review]
